code/game/objects/structures/safe.dm

```diff
@@ -68,22 +87,64 @@ FLOOR SAFES
 
 /obj/structure/safe/update_icon()
 	if(open)
-		icon_state = "[initial(icon_state)]-open"
+		if(broken)
+			icon_state = "[initial(icon_state)]-open-broken"
+		else
+			icon_state = "[initial(icon_state)]-open"
 	else
-		icon_state = initial(icon_state)
+		if(broken)
+			icon_state = "[initial(icon_state)]-broken"
+		else
+			icon_state = initial(icon_state)
 
+	cut_overlay(drill_overlay)
+	if(istype(drill, /obj/item/thermal_drill))
+		var/drill_icon = istype(drill, /obj/item/thermal_drill/diamond_drill) ? "d" : "h"
+		var/state = "[initial(icon_state)]_[drill_icon]-drill-[isprocessing ? "on" : "off"]"
+		drill_overlay = image(icon = 'icons/effects/drill.dmi', icon_state = state, pixel_x = drill_x_offset, pixel_y = drill_y_offset)
+		add_overlay(drill_overlay)
 
 /obj/structure/safe/attack_hand(mob/user as mob)
-	user.set_machine(src)
-	var/dat = "<center>"
-	dat += "<a href='?src=\ref[src];open=1'>[open ? "Close" : "Open"] [src]</a> | <a href='?src=\ref[src];decrement=1'>-</a> [dial * 5] <a href='?src=\ref[src];increment=1'>+</a>"
-	if(open)
-		dat += "<table>"
-		for(var/i = contents.len, i>=1, i--)
-			var/obj/item/P = contents[i]
-			dat += "<tr><td><a href='?src=\ref[src];retrieve=\ref[P]'>[P.name]</a></td></tr>"
-		dat += "</table></center>"
-	user << browse("<html><head><title>[name]</title></head><body>[dat]</body></html>", "window=safe;size=350x300")
+	if(drill)
+		switch(alert("What would you like to do?", "Thermal Drill", "Turn [isprocessing ? "Off" : "On"]", "Remove Drill", "Cancel"))
+			if("Turn On")
+				if(!drill || isprocessing)
+					return
+				if(broken)
+					to_chat(user, SPAN_WARNING("\The [src] is already broken open!"))
+					return
+				if(do_after(user, 2 SECONDS))
+					last_drill_time = world.time
+					drill.soundloop.start()
+					START_PROCESSING(SSprocessing, src)
+					update_icon()
```

Comment:
These loops need to check conditions again when the loop is complete. The drill might break or get removed in that timeframe.